File: bot/api/routes/bots.py
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query

from bot.api.dependencies import get_current_user_id
from bot.api.models import (
    BotCreate,
    BotResponse,
    BotUpdate,
    DecisionResponse,
    TradeResponse,
)
from bot.db.supabase import get_supabase

router = APIRouter()


def _get_bot_or_404(supabase, bot_id: str, user_id: str) -> dict:
    """Fetch a bot row and verify it belongs to the requesting user."""
    result = (
        supabase.table("trading_bots")
        .select("*")
        .eq("id", bot_id)
        .single()
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="Bot not found")
    bot = result.data
    if bot.get("user_id") != user_id:
        raise HTTPException(status_code=403, detail="Not authorised to access this bot")
    return bot


# ---------------------------------------------------------------------------
# List bots
# ---------------------------------------------------------------------------

@router.get("/", response_model=list[BotResponse])
async def list_bots(user_id: str = Depends(get_current_user_id)):
    supabase = get_supabase()
    result = (
        supabase.table("trading_bots")
        .select("*")
        .eq("user_id", user_id)
        .order("created_at", desc=True)
        .execute()
    )
    return result.data or []


# ---------------------------------------------------------------------------
# Create bot
# ---------------------------------------------------------------------------

@router.post("/", response_model=BotResponse, status_code=201)
async def create_bot(
    body: BotCreate,
    user_id: str = Depends(get_current_user_id),
):
    supabase = get_supabase()

    # Enforce handle uniqueness across the platform
    existing = (
        supabase.table("trading_bots")
        .select("id")
        .eq("handle", body.handle)
        .execute()
    )
    if existing.data:
        raise HTTPException(
            status_code=409, detail=f"Handle '{body.handle}' is already taken"
        )

    now = datetime.utcnow().isoformat()
    bot_id = str(uuid.uuid4())

    row = {
        "id": bot_id,
        "user_id": user_id,
        "handle": body.handle,
        "name": body.name,
        "model": body.model,
        "llm_provider": body.llm_provider,
        "symbol": body.symbol,
        "timeframes": body.timeframes,
        "system_prompt": body.system_prompt,
        "risk_mode": body.risk_mode,
        "max_position_usd": body.max_position_usd,
        "mode": body.mode,
        "hl_wallet_address": body.hl_wallet_address,
        "hl_private_key": body.hl_private_key,
        "status": "stopped",
        "equity_usd": None,
        "initial_equity_usd": 100.0,
        "total_trades": 0,
        "total_prompts": 0,
        "created_at": now,
        "updated_at": now,
    }

    result = supabase.table("trading_bots").insert(row).execute()
    if not result.data:
        raise HTTPException(status_code=500, detail="Failed to create bot")
    return result.data[0]


# ---------------------------------------------------------------------------
# Get single bot
# ---------------------------------------------------------------------------

@router.get("/{bot_id}", response_model=BotResponse)
async def get_bot(bot_id: str, user_id: str = Depends(get_current_user_id)):
    supabase = get_supabase()
    return _get_bot_or_404(supabase, bot_id, user_id)


# ---------------------------------------------------------------------------
# Update bot settings
# ---------------------------------------------------------------------------

@router.put("/{bot_id}", response_model=BotResponse)
async def update_bot(
    bot_id: str,
    body: BotUpdate,
    user_id: str = Depends(get_current_user_id),
):
    supabase = get_supabase()
    bot = _get_bot_or_404(supabase, bot_id, user_id)

    if bot.get("status") == "running":
        raise HTTPException(
            status_code=409,
            detail="Cannot update settings while bot is running. Stop it first.",
        )

    updates: dict = {
        k: v for k, v in body.model_dump(exclude_none=True).items()
    }
    if not updates:
        return bot

    updates["updated_at"] = datetime.utcnow().isoformat()
    result = (
        supabase.table("trading_bots")
        .update(updates)
        .eq("id", bot_id)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=500, detail="Update failed")
    return result.data[0]


# ---------------------------------------------------------------------------
# Delete bot
# ---------------------------------------------------------------------------

@router.delete("/{bot_id}", status_code=204)
async def delete_bot(bot_id: str, user_id: str = Depends(get_current_user_id)):
    supabase = get_supabase()
    bot = _get_bot_or_404(supabase, bot_id, user_id)

    if bot.get("status") == "running":
        raise HTTPException(
            status_code=409,
            detail="Stop the bot before deleting it.",
        )

    supabase.table("trading_bots").delete().eq("id", bot_id).execute()


# ---------------------------------------------------------------------------
# Start bot
# ---------------------------------------------------------------------------

@router.post("/{bot_id}/start", response_model=BotResponse)
async def start_bot(bot_id: str, user_id: str = Depends(get_current_user_id)):
    supabase = get_supabase()
    _get_bot_or_404(supabase, bot_id, user_id)

    now = datetime.utcnow().isoformat()
    result = (
        supabase.table("trading_bots")
        .update({"status": "running", "updated_at": now})
        .eq("id", bot_id)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=500, detail="Failed to start bot")
    return result.data[0]


# ---------------------------------------------------------------------------
# Stop bot
# ---------------------------------------------------------------------------

@router.post("/{bot_id}/stop", response_model=BotResponse)
async def stop_bot(bot_id: str, user_id: str = Depends(get_current_user_id)):
    supabase = get_supabase()
    _get_bot_or_404(supabase, bot_id, user_id)

    now = datetime.utcnow().isoformat()
    result = (
        supabase.table("trading_bots")
        .update({"status": "stopped", "updated_at": now})
        .eq("id", bot_id)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=500, detail="Failed to stop bot")
    return result.data[0]


# ---------------------------------------------------------------------------
# Bot decisions (paginated)
# ---------------------------------------------------------------------------

@router.get("/{bot_id}/decisions", response_model=list[DecisionResponse])
async def get_decisions(
    bot_id: str,
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
    user_id: str = Depends(get_current_user_id),
):
    supabase = get_supabase()
    _get_bot_or_404(supabase, bot_id, user_id)

    offset = (page - 1) * limit
    result = (
        supabase.table("bot_decisions")
        .select("*")
        .eq("bot_id", bot_id)
        .order("created_at", desc=True)
        .range(offset, offset + limit - 1)
        .execute()
    )
    return result.data or []


# ---------------------------------------------------------------------------
# Bot trades (paginated)
# ---------------------------------------------------------------------------

@router.get("/{bot_id}/trades", response_model=list[TradeResponse])
async def get_trades(
    bot_id: str,
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
    user_id: str = Depends(get_current_user_id),
):
    supabase = get_supabase()
    _get_bot_or_404(supabase, bot_id, user_id)

    offset = (page - 1) * limit
    result = (
        supabase.table("bot_trades")
        .select("*")
        .eq("bot_id", bot_id)
        .order("opened_at", desc=True)
        .range(offset, offset + limit - 1)
        .execute()
    )
    return result.data or []

File: api/src/routes/bots.ts
import { Router, type Request } from "express";
import { z } from "zod";
import { requireAuth, type AuthenticatedRequest } from "../middleware/auth.js";
import { supabase } from "../services/supabase.js";

export const botsRouter = Router();

botsRouter.use(requireAuth);

function auth(req: Request): AuthenticatedRequest {
  return req as unknown as AuthenticatedRequest;
}

const CreateBotSchema = z.object({
  handle: z
    .string()
    .min(2)
    .max(32)
    .regex(/^[a-zA-Z0-9_-]+$/, "Handle must be alphanumeric with _ or -"),
  name: z.string().min(1).max(64),
  model: z.string().min(1).default("claude-haiku-4-5-20251001"),
  symbol: z.string().min(1).max(20).default("BTC"),
  timeframes: z.array(z.string()).min(1).default(["30m", "4h", "1d"]),
  riskMode: z.enum(["conservative", "moderate", "aggressive"]).default("moderate"),
  maxPositionUsd: z.number().positive().max(1_000_000).default(100),
  systemPrompt: z.string().max(4000).optional(),
  hlWalletAddress: z.string().optional(),
  mode: z.enum(["server", "client"]).default("server"),
});

const UpdateBotSchema = CreateBotSchema.partial().omit({ handle: true });

// GET /api/bots
botsRouter.get("/", async (req, res) => {
  const { userId } = auth(req);

  const { data, error } = await supabase
    .from("trading_bots")
    .select("*")
    .eq("user_id", userId)
    .order("created_at", { ascending: false });

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }

  res.json(data);
});

// POST /api/bots
botsRouter.post("/", async (req, res) => {
  const { userId } = auth(req);

  const parsed = CreateBotSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.flatten() });
    return;
  }

  const {
    handle,
    name,
    model,
    symbol,
    timeframes,
    riskMode,
    maxPositionUsd,
    systemPrompt,
    hlWalletAddress,
    mode,
  } = parsed.data;

  const { data, error } = await supabase
    .from("trading_bots")
    .insert({
      user_id: userId,
      handle,
      name,
      model,
      symbol,
      timeframes,
      risk_mode: riskMode,
      max_position_usd: maxPositionUsd,
      system_prompt: systemPrompt ?? null,
      hl_wallet_address: hlWalletAddress ?? null,
      mode,
    })
    .select()
    .single();

  if (error) {
    if (error.code === "23505") {
      res.status(409).json({ error: "Handle already taken" });
      return;
    }
    res.status(500).json({ error: error.message });
    return;
  }

  res.status(201).json(data);
});

// GET /api/bots/:id
botsRouter.get("/:id", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const { data: bot, error: botError } = await supabase
    .from("trading_bots")
    .select("*")
    .eq("id", id)
    .eq("user_id", userId)
    .single();

  if (botError || !bot) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }

  const { data: decisions } = await supabase
    .from("bot_decisions")
    .select("*")
    .eq("bot_id", id)
    .order("created_at", { ascending: false })
    .limit(10);

  res.json({ ...bot, recentDecisions: decisions ?? [] });
});

// PUT /api/bots/:id
botsRouter.put("/:id", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const parsed = UpdateBotSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.flatten() });
    return;
  }

  const updates: Record<string, unknown> = {};
  const d = parsed.data;

  if (d.name !== undefined) updates.name = d.name;
  if (d.model !== undefined) updates.model = d.model;
  if (d.symbol !== undefined) updates.symbol = d.symbol;
  if (d.timeframes !== undefined) updates.timeframes = d.timeframes;
  if (d.riskMode !== undefined) updates.risk_mode = d.riskMode;
  if (d.maxPositionUsd !== undefined) updates.max_position_usd = d.maxPositionUsd;
  if (d.systemPrompt !== undefined) updates.system_prompt = d.systemPrompt;
  if (d.hlWalletAddress !== undefined) updates.hl_wallet_address = d.hlWalletAddress;
  if (d.mode !== undefined) updates.mode = d.mode;

  if (Object.keys(updates).length === 0) {
    res.status(400).json({ error: "No valid fields to update" });
    return;
  }

  const { data, error } = await supabase
    .from("trading_bots")
    .update(updates)
    .eq("id", id)
    .eq("user_id", userId)
    .select()
    .single();

  if (error || !data) {
    res.status(404).json({ error: "Bot not found or update failed" });
    return;
  }

  res.json(data);
});

// DELETE /api/bots/:id
botsRouter.delete("/:id", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const { error } = await supabase
    .from("trading_bots")
    .delete()
    .eq("id", id)
    .eq("user_id", userId);

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }

  res.json({ ok: true });
});

// POST /api/bots/:id/start
botsRouter.post("/:id/start", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const { data, error } = await supabase
    .from("trading_bots")
    .update({ status: "running" })
    .eq("id", id)
    .eq("user_id", userId)
    .select()
    .single();

  if (error || !data) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }

  res.json({ ok: true, status: (data as { status: string }).status });
});

// POST /api/bots/:id/stop
botsRouter.post("/:id/stop", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const { data, error } = await supabase
    .from("trading_bots")
    .update({ status: "stopped" })
    .eq("id", id)
    .eq("user_id", userId)
    .select()
    .single();

  if (error || !data) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }

  res.json({ ok: true, status: (data as { status: string }).status });
});

// GET /api/bots/:id/decisions
botsRouter.get("/:id/decisions", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const page = Math.max(1, parseInt(String(req.query.page ?? "1"), 10));
  const limit = Math.min(100, Math.max(1, parseInt(String(req.query.limit ?? "20"), 10)));
  const offset = (page - 1) * limit;

  const { data: bot } = await supabase
    .from("trading_bots")
    .select("id")
    .eq("id", id)
    .eq("user_id", userId)
    .single();

  if (!bot) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }

  const { data, error, count } = await supabase
    .from("bot_decisions")
    .select("*", { count: "exact" })
    .eq("bot_id", id)
    .order("created_at", { ascending: false })
    .range(offset, offset + limit - 1);

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }

  res.json({ data: data ?? [], total: count ?? 0, page, limit });
});

// GET /api/bots/:id/trades
botsRouter.get("/:id/trades", async (req, res) => {
  const { userId } = auth(req);
  const { id } = req.params;

  const page = Math.max(1, parseInt(String(req.query.page ?? "1"), 10));
  const limit = Math.min(100, Math.max(1, parseInt(String(req.query.limit ?? "20"), 10)));
  const offset = (page - 1) * limit;

  const { data: bot } = await supabase
    .from("trading_bots")
    .select("id")
    .eq("id", id)
    .eq("user_id", userId)
    .single();

  if (!bot) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }

  const { data, error, count } = await supabase
    .from("bot_trades")
    .select("*", { count: "exact" })
    .eq("bot_id", id)
    .order("opened_at", { ascending: false })
    .range(offset, offset + limit - 1);

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }

  res.json({ data: data ?? [], total: count ?? 0, page, limit });
});
